### PR TITLE
use provider specific parameters

### PR DESCRIPTION
**Pull request by dbraquart**

### CODE REVIEW

File: src/main/java/org/gridsuite/loadflow/server/dto/ParameterInfos.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Getter;
+import java.util.List;
+
+/**
+ * @author David Braquart <david.braquart@rte-france.com>
+ */
+@Getter
+@AllArgsConstructor
+@Builder
+public class ParameterInfos {
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/loadflow/server/dto/ParameterInfos.java:21]: com.powsybl.commons.parameters.Parameter is already serializable (adding the correct jackson module) and could be used instead

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java [hunk 2]
```diff
@@ -68,12 +71,25 @@ private Network getNetwork(UUID networkUuid) {
         }
     }
 
-    LoadFlowResult run(UUID networkUuid, String variantId, List<UUID> otherNetworksUuid, LoadFlowParameters parameters,
+    private LoadFlowParameters buildLoadFlowParameters(LoadFlowParametersInfos loadflowParams, String provider) {
+        LoadFlowParameters params = loadflowParams == null || loadflowParams.getCommonParameters() == null ?
+             new LoadFlowParameters() : loadflowParams.getCommonParameters();
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java:76]: LoadFlowParameters.load() would be better than new LoadFlowParameters() to allows loading default values specific to some deployment

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java [hunk 3]
```diff
@@ -77,4 +71,13 @@ public ResponseEntity<List<String>> getProviders() {
     public ResponseEntity<String> getDefaultLoadflowProvider() {
         return ResponseEntity.ok().body(loadFlowService.getDefaultProvider());
     }
+
+    @GetMapping(value = "/specific-parameters")
+    @Operation(summary = "Get all existing loadflow specific parameters for a given provider, or for all of them")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The loadflow model-specific parameters")})
+    public ResponseEntity<Map<String, List< com.powsybl.commons.parameters.Parameter>>> getSpecificLoadflowParameters(
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java:78]: why is it fully qualified?

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java [hunk 4]
```diff
@@ -77,4 +71,13 @@ public ResponseEntity<List<String>> getProviders() {
     public ResponseEntity<String> getDefaultLoadflowProvider() {
         return ResponseEntity.ok().body(loadFlowService.getDefaultProvider());
     }
+
+    @GetMapping(value = "/specific-parameters")
+    @Operation(summary = "Get all existing loadflow specific parameters for a given provider, or for all of them")
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java:76]: Ok directly discussed, from the front you directly create the UI for all possible providers

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java [hunk 5]
```diff
@@ -137,4 +152,15 @@ List<String> getProviders() {
     public String getDefaultProvider() {
         return defaultProvider;
     }
+
+    public Map<String, List< Parameter>> getSpecificLoadFlowParameters(String providerName) {
+        return LoadFlowProvider.findAll().stream()
+                .filter(provider -> providerName == null || provider.getName().equals(providerName))
```
[the remark this answers is not in the document]
Comment (dbraquart) [reply src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java:158]: As we discussed directly, I need to retrieve specific parameters for all providers at a time. As documented in the Rest API, we can do that omitting the provider name (it is null in this function).

File: src/test/java/org/gridsuite/loadflow/server/LoadFlowTest.java [hunk 6]
```diff
@@ -119,20 +122,16 @@ public void test() throws Exception {
...
-
-        ByteArrayOutputStream stream = new ByteArrayOutputStream();
-        JsonLoadFlowParameters.write(params, stream);
-        String paramsString = stream.toString();
+        LoadFlowParametersInfos fullParams = LoadFlowParametersInfos.builder()
+                .commonParameters(new LoadFlowParameters())
+                .specificParameters(Map.of("SlackBusSelectionMode", "MOST_MESHED"))
+                .build();
+        String paramsString = objectWriter.writeValueAsString(fullParams);
+        paramsString = paramsString.replace("\"extensions\" : [ ],", "");
```
[the remark this answers is not in the document]
Comment (dbraquart) [reply src/test/java/org/gridsuite/loadflow/server/LoadFlowTest.java:130]: good to know, I'll try your solution

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java (hunk 3 above)
Comment (dbraquart) [reply src/main/java/org/gridsuite/loadflow/server/LoadFlowController.java:78]: I have to use both io.swagger.v3.oas.annotations.Parameter and com.powsybl.commons.parameters.Parameter.
I dont know how to handle this differently.

File: src/test/java/org/gridsuite/loadflow/server/LoadFlowTest.java [hunk 7]
```diff
@@ -151,6 +150,34 @@ public void test() throws Exception {
...
 
+    private void simpleRunWithLFParams(LoadFlowParameters lfParams, Map<String, String> specificParams) throws Exception {
+        given(networkStoreService.getNetwork(TEST_NETWORK_ID, PreloadingStrategy.ALL_COLLECTIONS_NEEDED_FOR_BUS_VIEW)).willReturn(createNetwork(true));
+
+        LoadFlowParametersInfos fullParams = LoadFlowParametersInfos.builder()
+                .commonParameters(lfParams)
+                .specificParameters(specificParams)
+                .build();
+        String paramsString = objectWriter.writeValueAsString(fullParams);
+        paramsString = paramsString.replace("\"extensions\" : [ ],", "");
```
[the remark this answers is not in the document]
Comment (dbraquart) [reply src/test/java/org/gridsuite/loadflow/server/LoadFlowTest.java:161]: yes done

File: src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java [hunk 8]
```diff
@@ -46,7 +49,7 @@
 @Service
 class LoadFlowService {
 
-    @Value("${report-server.base-uri:http://report-server}")
+    @Value("${gridsuite.services.report-server.base-uri:http://report-server}")
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/loadflow/server/LoadFlowService.java:52]: add the "/" at the end to be homogeneous

**Review by etiennehomer — APPROVED** [review]

**Review by geofjamg — APPROVED** [review]

**Pull request merged**